File: LeetCode/Hao/260SingleNumberIII.java
public class Solution {
    public int[] singleNumber(int[] nums) {
        int x, y;
        x = y = 0;
        
        for (int i = 0; i < nums.length; i ++) {
            x ^= nums[i];
        }
        
        int lowerbit = (x & (-x));
        
        y = x; 
        
        for (int i = 0; i < nums.length; i ++) {
            if ((lowerbit & nums[i]) != 0) {
                x ^= nums[i];
            }
        }
        
        y ^= x;
        
        return new int[]{x, y};
    }
}

File: LeetCode/Liang/260SingleNumberIII.cpp
class Solution {
public:
    vector<int> singleNumber(vector<int>& nums) {
        assert(nums.size() >= 2);
        int aXORb = 0; // Use 0 is fine.
        for (int i = 0; i < nums.size(); ++i) {
            aXORb ^= nums.at(i);
        }
        int lastDiff = (aXORb & (aXORb - 1)) ^ aXORb;
        int first = 0, second = 0;
        for (int i = 0; i < nums.size(); ++i) {
            if (lastDiff & nums.at(i)) {
                first ^= nums.at(i);
            } else {
                second ^= nums.at(i);
            }
        }
        return vector<int>{first, second};
    }
};
